lib/task_listener.js

```diff
@@ -13,166 +13,275 @@ var request = require('superagent-promise');
 
 var Task = require('./task');
 var EventEmitter = require('events').EventEmitter;
+var util = require('util');
+var TaskQueue = require('./queueservice');
 
 /**
 @param {Configuration} config for worker.
 */
-function TaskListener(runtime) {
-  this.runtime = runtime;
-  EventEmitter.call(this);
-}
+export default class TaskListener extends EventEmitter {
+  constructor(runtime) {
+    this.pending = 0;
+    this.runtime = runtime;
+    this.capacity = runtime.capacity;
+    this.runningTasks = [];
+    this.taskQueue = new TaskQueue(this.runtime);
+    this.taskPollInterval = this.runtime.taskQueue.pollInterval;
+    this.claimConfig = {
+      workerId: this.runtime.workerId,
+      workerGroup: this.runtime.workerGroup
+    };
+    super();
+  }
 
-TaskListener.prototype = {
-  __proto__: EventEmitter.prototype,
+  listenForShutdowns() {
+    // If node will be shutdown, stop consuming events.
+    if (this.runtime.shutdownManager) {
+      this.runtime.shutdownManager.once(
+        'nodeTermination', () => {
+          debug('nodeterm');
+          async () => {
+            await this.pause();
+            for(let task of this.runningTasks) {
+              task.abort('worker-shutdown');
+            }
+          }();
+        }.bind(this)
+      );
+    }
+  }
 
-  /**
-  Number of running tasks...
-  */
-  pending: 0,
+  async cancelTask(message) {
+    var runId = message.payload.runId;
+    var reason = message.payload.status.runs[runId].reasonResolved;
+    if (reason !== 'canceled') return;
 
-  connect: function* () {
-    this.runtime.log('listener connect');
+    var taskId = message.payload.status.taskId;
+    var task = this.runningTasks.find(
+      (task) => { return (task.status.taskId === taskId && task.runId === runId); }
+    );
 
-    var self = this;
-    var queue = this.runtime.queue;
+    if (!task) debug('task not found to cancel'); return;
 
-    // Share the queue between all workerTypes of the same provisioner.
-    var queueName;
-    if (this.runtime.createQueue) {
-      queueName =
-        QUEUE_PREFIX +
-        this.runtime.provisionerId + '/' + this.runtime.workerType;
-    }
+    this.runtime.log('cancelling task', {taskId: message.payload.status.taskId});
+    task.cancel(reason);
```

Comment:
I'm not sure you are aware of `PulseListener` semantics. The event handler may return a promise.
If the promise is rejected, `PulseListener` will reject the event and the event will happen again, second time it fails it'll be logged and discarded.

So if you wanted to you could return a promise from `task.cancel()`. Though I'm not sure this is always desired. So maybe you have the right semantics :)